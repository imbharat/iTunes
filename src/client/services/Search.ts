import axios from "axios"
import { Filter } from "../../redux/Songs/Types";
import CONFIG from "../Configuration";

export const fetchData = async (filter: Filter) => {
    const data = await axios.get(`${CONFIG['DEV'].URL}/api/search`, {
        params: { ...filter }
    });
    return data;
}