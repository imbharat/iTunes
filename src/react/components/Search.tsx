import React, { useCallback, useEffect, useState } from 'react'
import { TextField, Button, Dialog, DialogActions, DialogContent, 
    Select, MenuItem, DialogTitle, FormControl,
    FormControlLabel, Radio, RadioGroup, FormLabel } from '@mui/material';
import useDebounce from '../../client/customHooks/useDebounce';
import { useAppDispatch, useAppSelector } from '../../redux/ReduxHooks';
import { updateFilterAndLoadSongs } from '../../redux/Songs/SongsActions';
import { Songs } from '../../redux/Songs/Types';
import { StyledSearch } from '../styled-components/StyledSearch';

function Search() {
    const [firstRender, setFirstRender] = useState(true);
    const [keywords, setKeywords] = useState('');
    const [searchKeywords, setSearchKeywords] = useState('');
    const [open, setOpen] = React.useState(false);

    const initialFilters = {
        country: 'US',
        media: 'all',
        entity: '',
        attribute: '',
        lang: 'en_us'
    }
    const [filters, setFilters] = useState(initialFilters)
    
    const debouncedKeywords = useCallback(
        useDebounce(setSearchKeywords, 1000), []
    );
    
    const updateQuery = (e: React.ChangeEvent<HTMLInputElement>) => {
        setKeywords(e.target.value);
        debouncedKeywords(e.target.value);
    }

    const dispatch = useAppDispatch();
    const filter = useAppSelector((state: Songs) => state.filter)

    const handleClose = (search: boolean) => {
        setOpen(false);
        search && dispatch(updateFilterAndLoadSongs({
            ...filter,
            page: 1,
            ...filters
        }, true));
        setFilters(initialFilters);
    };

    useEffect(() => {
        if(!firstRender){
            dispatch(updateFilterAndLoadSongs({
                ...filter,
                term: searchKeywords[0],
                page: 1
            }, true))
        }
        setFirstRender(false);
        return () => {
            //write cleanup code here
        }   
    }, [searchKeywords]);
    
    return (
        <StyledSearch>
            <TextField 
                placeholder='Type Keywords'
                data-testid='search'
                value={keywords}
                onChange={updateQuery}
                variant="outlined"
            />
            <Button
                data-testid='advanced-search'
                variant="contained"
                onClick={() => setOpen(true)}
            >
                Advanced Search
            </Button>

            {
                open && <Dialog data-testid='advanced-dialog' open={open} onClose={() => handleClose(false)}>
                    <DialogTitle>Advanced Search</DialogTitle>
                    <DialogContent className='advSearch'>
                        <TextField
                            autoFocus
                            margin="dense"
                            id="country"
                            label="Country"
                            variant="standard"
                            style={{
                                marginRight: '5rem'
                            }}
                            value={filters.country}
                            onChange={(e) => setFilters({
                                ...filters,
                                country: e.target.value
                            })}
                        />
                        <FormControl>
                            <FormLabel id="demo-lang-radio-buttons-group-label">Language</FormLabel>
                            <RadioGroup
                                id="lang"
                                row
                                onChange={(e) => setFilters({
                                    ...filters,
                                    lang: e.target.value
                                })}
                                name="lang-radio-buttons-group"
                            >
                                <FormControlLabel value="en_us" control={<Radio />} label="EN" />
                                <FormControlLabel value="ja_jp" control={<Radio />} label="JP" />
                            </RadioGroup>
                        </FormControl>
                        <Select
                            id="media"
                            value={filters.media}
                            label="Media Type"
                            fullWidth
                            onChange={(e) => setFilters({
                                ...filters,
                                media: e.target.value
                            })}
                        >
                            <MenuItem value='all'>All</MenuItem>
                            <MenuItem value='movie'>Movie</MenuItem>
                            <MenuItem value='podcast'>Podcast</MenuItem>
                            <MenuItem value='music'>Music</MenuItem>
                            <MenuItem value='musicVideo'>Music Video</MenuItem>
                            <MenuItem value='audiobook'>Audio Book</MenuItem>
                            <MenuItem value='shortFilm'>Short Film</MenuItem>
                            <MenuItem value='tvShow'>TV Show</MenuItem>
                            <MenuItem value='software'>Software</MenuItem>
                            <MenuItem value='ebook'>E-Book</MenuItem>
                        </Select>
                        {/* <TextField
                            margin="dense"
                            id="entity"
                            label="Entity"
                            type="email"
                            fullWidth
                            variant="standard"
                        />
                        <TextField
                            margin="dense"
                            id="attribute"
                            label="Attribute"
                            type="email"
                            fullWidth
                            variant="standard"
                        /> */}
                    </DialogContent>
                    <DialogActions>
                        <Button onClick={() => handleClose(false)}>Cancel</Button>
                        <Button onClick={() => handleClose(true)}>Search</Button>
                    </DialogActions>
                </Dialog>
            }

        </StyledSearch>
    )
}

export default Search