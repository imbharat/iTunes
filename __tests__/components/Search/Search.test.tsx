import React from 'react';
import renderer from 'react-test-renderer';
import Search from '../../../src/react/components/Search';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';

const middlewares = [thunk]
const mockStore = configureStore(middlewares);

describe("Search Component Tests", () => {
    let store: any;
    let component: any;
    beforeEach(() => {

      store = mockStore({
        loading: false,
        songs: [],
        filter: {
            term: '',
            country: '',
            media: '',
            entity: '',
            attribute: '',
            limit: 50,
            lang: '',
            page: 0
        },
        isLast: true,
        error: ''
      });

      store.dispatch = jest.fn();

      component = renderer.create(
        <Provider store={store}>
            <Search />
        </Provider>
      );
    });

    it('search component rendering', () => {
      //act
      const tree = component.toJSON();
      //assert
      expect(tree).toMatchSnapshot();
    });

    it('keywords change dispatch action', async () => {
        //act
        renderer.act(() => {
            render(
                <Provider store={store}>
                    <Search />
                </Provider>
            );
        });
        const searchElement = screen.getByTestId('search').querySelector('input')!;
        renderer.act(() => {
            fireEvent.change(searchElement, {target: {value: 'Jackson'}});
        });
        await sleep(2000);
        //assert
        expect(store.dispatch).toHaveBeenCalledTimes(1);
    });

    it('advanced search button exists in search bar', () => {
        //act
        renderer.act(() => {
          render(
              <Provider store={store}>
                  <Search />
              </Provider>
          );
        });
        const searchButton = screen.getByTestId('advanced-search').parentElement?.querySelector('button')!;
        expect(searchButton).not.toBeNull();
    });

    it('advanced search button click opens dialog', () => {
        //act
        renderer.act(() => {
          render(
              <Provider store={store}>
                  <Search />
              </Provider>
          );
        });
        const searchButton = screen.getByTestId('advanced-search').parentElement?.querySelector('button')!;
        renderer.act(() => {
            fireEvent.click(searchButton);
        });
        const dialog = screen.getByTestId('advanced-dialog');
        expect(dialog).not.toBeNull();
    });
});

const sleep = function (timeout: number) {
    return new Promise(function (resolve, reject) {
        setTimeout(function () {
            resolve('time passed');
        }, timeout);
    });
};