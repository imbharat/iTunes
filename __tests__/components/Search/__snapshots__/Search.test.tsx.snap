// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component Tests search component rendering 1`] = `
<div
  className="sc-bczRLJ kEAijm"
>
  <div
    className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    data-testid="search"
  >
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r0:"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Type Keywords"
        required={false}
        type="text"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-hdw1oc"
        >
          <span
            className="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
    data-testid="advanced-search"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Advanced Search
  </button>
</div>
`;
